domain-management/src/main/java/org/jboss/as/domain/management/security/FileKeyManagerService.java

```diff
@@ -117,45 +141,117 @@ public void setAlias(String alias) {
-                public void pathModelEvent(PathEventContext eventContext, String name) {
-                    if (eventContext.isResourceServiceRestartAllowed() == false) {
-                        eventContext.reloadRequired();
-                    }
+    public void stop(StopContext context) {
+        super.stop(context);
+        keyStore = null;
+    }
+
+    @Override
+    protected boolean isLazy() {
+        return keyStore == null;
+    }
+
+    @Override
+    protected KeyStore loadKeyStore(boolean startup) {
+        try {
+            if (keyStore != null) {
+                if (keyStore.isModified()) {
+                    keyStore.load();
                 }
+                return keyStore.getKeyStore();
+            }
+            String file = path;
+            if (relativeTo != null) {
+                PathManager pm = pathManager.getValue();
+
+                file = pm.resolveRelativePathEntry(file, relativeTo);
+                pm.registerCallback(relativeTo, new Callback() {
 
-                @Override
-                public void pathEvent(Event event, PathEntry pathEntry) {
-                    // Service dependencies should trigger a stop and start.
+                    @Override
+                    public void pathModelEvent(PathEventContext eventContext, String name) {
+                        if (!eventContext.isResourceServiceRestartAllowed()) {
+                            eventContext.reloadRequired();
+                        }
+                    }
+
+                    @Override
+                    public void pathEvent(Event event, PathEntry pathEntry) {
+                        // Service dependencies should trigger a stop and start.
+                    }
+                }, Event.REMOVED, Event.UPDATED);
+            }
+            Path path = Paths.get(file);
+            if (!Files.exists(path) && autoGenerateCertHostName != null) {
+                if(startup) {
+                    DomainManagementLogger.SECURITY_LOGGER.keystoreWillBeCreated(file, autoGenerateCertHostName);
+                    return null;
+                } else {
+                    generateFileKeyStore(path);
                 }
-            }, Event.REMOVED, Event.UPDATED);
-        }
+            }
 
-        keyStore = FileKeystore.newKeyStore(provider, file, keystorePassword, keyPassword, alias);
-        keyStore.load();
+            keyStore = FileKeystore.newKeyStore(provider, file, keystorePassword, keyPassword, alias);
+            keyStore.load();
 
-        super.start(context);
+            return keyStore.getKeyStore();
+        } catch (StartException e) {
+            throw new IllegalStateException(e);
+        }
     }
 
-    @Override
-    public void stop(StopContext context) {
-        super.stop(context);
-        keyStore = null;
+    private void generateFileKeyStore(Path path) {
+        try {
+            DomainManagementLogger.SECURITY_LOGGER.keystoreWillBeCreated(path.toString());
```

Comment:
I think in addition to logging that this is being created it would be beneficial to log information about the self signed certificate - at least this way the users have a chance to verify the certificate is the one generated by the server.